port/zos390/omriarv64.mc

```diff
@@ -0,0 +1,417 @@
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "omriarv64.h"
+
+#pragma prolog(omrallocate_1M_fixed_pages,"MYPROLOG")
+#pragma epilog(omrallocate_1M_fixed_pages,"MYEPILOG")
+
+__asm(" IARV64 PLISTVER=MAX,MF=(L,LGETSTOR)":"DS"(lgetstor));
+
+/*
+ * Allocate 1MB fixed pages using IARV64 system macro.
+ * Memory allocated is freed using omrfree_memory_above_bar().
+ *
+ * @params[in] numMBSegments Number of 1MB segments to be allocated
+ * @params[in] userExtendedPrivateAreaMemoryType capability of OS: 0 - general, 1 - 2G-32G range, 2 - 2G-64G range
+ *
+ * @return pointer to memory allocated, NULL on failure.
+ */
+void * omrallocate_1M_fixed_pages(int *numMBSegments, int *userExtendedPrivateAreaMemoryType, const char * ttkn, void* startAddress) {
+	long segments;
+	long origin;
+	long useMemoryType = *userExtendedPrivateAreaMemoryType;
+	int  iarv64_rc = 0;
+	__asm(" IARV64 PLISTVER=MAX,MF=(L,LGETSTOR)":"DS"(wgetstor));
+
+	segments = *numMBSegments;
+	wgetstor = lgetstor;
+
+	if (startAddress == (void*)0) {
+		switch (useMemoryType) {
+        case ZOS64_VMEM_ABOVE_BAR_GENERAL:
+			__asm(" IARV64 REQUEST=GETSTOR,COND=YES,SADMP=NO,CONTROL=UNAUTH,PAGEFRAMESIZE=1MEG,"\
+					"SEGMENTS=(%2),ORIGIN=(%1),TTOKEN=(%4),RETCODE=%0,MF=(E,(%3))"\
+					::"m"(iarv64_rc),"r"(&origin),"r"(&segments),"r"(&wgetstor),"r"(ttkn));
+			break;
+		case ZOS64_VMEM_2_TO_32G:
+			__asm(" IARV64 REQUEST=GETSTOR,COND=YES,SADMP=NO,USE2GTO32G=YES,"\
+					"CONTROL=UNAUTH,PAGEFRAMESIZE=1MEG,"\
+					"SEGMENTS=(%2),ORIGIN=(%1),TTOKEN=(%4),RETCODE=%0,MF=(E,(%3))"\
+					::"m"(iarv64_rc),"r"(&origin),"r"(&segments),"r"(&wgetstor),"r"(ttkn));
+			break;
+		case ZOS64_VMEM_2_TO_64G:
+			__asm(" IARV64 REQUEST=GETSTOR,COND=YES,SADMP=NO,USE2GTO64G=YES,"\
+					"CONTROL=UNAUTH,PAGEFRAMESIZE=1MEG,"\
+					"SEGMENTS=(%2),ORIGIN=(%1),TTOKEN=(%4),RETCODE=%0,MF=(E,(%3))"\
+					::"m"(iarv64_rc),"r"(&origin),"r"(&segments),"r"(&wgetstor),"r"(ttkn));
+			break;
+		}
+	} else {
+		switch (useMemoryType) {
+		case ZOS64_VMEM_ABOVE_BAR_GENERAL:
+			__asm(" IARV64 REQUEST=GETSTOR,COND=YES,SADMP=NO,CONTROL=UNAUTH,PAGEFRAMESIZE=1MEG,"\
+					"SEGMENTS=(%2),ORIGIN=(%1),TTOKEN=(%4),RETCODE=%0,MF=(E,(%3)),INORIGIN=(%5)"\
+					::"m"(iarv64_rc),"r"(&origin),"r"(&segments),"r"(&wgetstor),"r"(ttkn),"r"(&startAddress));
```

Comment:
What are the minimum machine requirements for compiling this? I assume 2.2 but if so, what APAR is required? I expect this change is updating the minimum machine requirements from 2.1

Reply:
Min build machine requirement is z/OS 2.2 and above. 

I'll update the APAR numbers once they are available. The new IARV64 underwent a change 1.5 month ago and I haven't asked the new number.

Reply:
The minimum requirements needs to be clarified, and OMR may need to accept a change to the minimum requirements, and will need to have them installed in order for this code to compile. Not sure if OMR has a document for minimum machine requirements, but if so it may need to be updated.

Reply:
Do the OMR machines have 2.2 plus the APAR installed?